cmd/snap-bootstrap/initramfs_systemd_mount.go

```diff
@@ -149,6 +156,23 @@ func doSystemdMountImpl(what, where string, opts *systemdMountOptions) error {
 	if opts.Private {
 		options = append(options, "private")
 	}
+	if opts.VerityHashDevice != "" && opts.VerityRootHash == "" {
+		return fmt.Errorf("cannot mount %q at %q: mount with dm-verity was requested but a root hash was not specified", what, where)
+	}
+	if opts.VerityRootHash != "" && opts.VerityHashDevice == "" {
+		return fmt.Errorf("cannot mount %q at %q: mount with dm-verity was requested but a hash device was not specified", what, where)
+	}
+	if opts.VerityHashOffset != 0 && (opts.VerityHashDevice == "" || opts.VerityRootHash == "") {
+		return fmt.Errorf("cannot mount %q at %q: mount with dm-verity was requested but a hash device and root hash were not specified", what, where)
+	}
+	if opts.VerityHashDevice != "" && opts.VerityRootHash != "" {
+		options = append(options, fmt.Sprintf("verity.roothash=%s", opts.VerityRootHash))
+		options = append(options, fmt.Sprintf("verity.hashdevice=%s", opts.VerityHashDevice))
```

Comment:
Do we need to possibly escape the hash device path? Is there any convention about this? What if the device has a space or a comma in the name?

Reply:
Thanks for catching this. I had already done escaping for the comma in the overlayfs PR (https://github.com/canonical/snapd/commit/9850c03f98fbf608b0a3a466cbeb924c0d3e1649) so I added the same here.

I added an extra test case for this but I have it commented out for now as the test code will need to use `SplitMountOptions()` (introduced in https://github.com/canonical/snapd/pull/14179/commits/f1171b9f262455992ff27b8fd637d23d04d3575f and later moved in https://github.com/canonical/snapd/pull/14179/commits/3477400a6ac0234f73cf45db30f9c569f1757631) to properly split the mount options. The code for `SplitMountOptions()` will be introduced as part of the overlayfs PR, so once that one is merged I will uncomment the test and remove the TODO comment.

Spaces don't seem to need escaping. These are options that get passed to `systemd-mount` when it's ran using Go's Exec so I think the shell escapes those for us. This is some example code I used to test this:
```bash
# First create the necessary folders under tmp
mkdir -p /tmp/overlay/lower\ 
mkdir -p /tmp/overlay/upper
mkdir -p /tmp/overlay/work
```

```Go
func (s *doSystemdMountSuite) TestSpace(c *C) {
	dirWithSpace := "/tmp/overlay/lower "
	args := []string{dirwithSpace, "/tmp/overlay/merged"}
	args = append(args, "--type=overlay")

	var options []string
	lowerdir := dirWithSpace
	upperdir := "/tmp/overlay/upper"
	workdir := "/tmp/overlay/work"
	options = append(options, fmt.Sprintf("lowerdir=%s", strings.ReplaceAll(lowerdir, ",", `\,`)))
	options = append(options, fmt.Sprintf("upperdir=%s", strings.ReplaceAll(upperdir, ",", `\,`)))
	options = append(options, fmt.Sprintf("workdir=%s", strings.ReplaceAll(workdir, ",", `\,`)))

	args = append(args, "--options="+strings.Join(options, ","))
	cmd := exec.Command("systemd-mount", args...)

	// Run the command and capture the output
	output, err := cmd.Output()

	// Check for errors
	if err != nil {
		log.Fatalf("Error running command: %v", err)
	}

	// Print the output
	fmt.Printf("Command output: %s\n", output)
}
```

Reply:
Spaces do need escaping. Look at the issues with p9 file system. If you don't escape spaces then /proc/self/mountinfo cannot be parsed correctly, as it has optional fields and options (the naming is beyond our control) and both are separated by spaces from each other (any number of optional fields and the singular "options" field).

Reply:
Interesting! Do we support this in snapd currently? I guess I could perform the same escaping for spaces as well in addition to the commas (and I guess I would also need to do that in the overlay PR). I am not aware of a more exact convention about this.

Reply:
I hope we don't have to use spaces but the point is that the options field is quite fragile so we should employ some escape logic for it. My advice is to look at the source code carefully: https://elixir.bootlin.com/linux/v6.10.9/source/fs/overlayfs/params.c#L189

The unfortunate reality is that it is a mess and literally every routine does something custom. In some places spaces are escaped, in others they are not.

Reply:
Added a commit for this to both this and the overlay PR.